src/render/idyll-display/components/author-tool-buttons.js

```diff
@@ -0,0 +1,333 @@
+      this.setState({
+        showCode: false
+      });
+    } else {
+      this._markup = this.getMarkup(this.props);
+      this.setState({ showCode: true });
+    }
+  }
+
+  handleClickProps() {
+    if (
+      this.context.activeComponent &&
+      this.context.activeComponent.id === this.props.idyllASTNode.id
+    ) {
+      this.context.setActiveComponent(null);
+    } else {
+      this.context.setActiveComponent({ ...this.props.idyllASTNode });
+    }
+  }
+
+  getMarkup(props) {
+    return AST.toMarkup({
+      id: -1,
+      type: 'component',
+      name: 'div',
+      children: [props.idyllASTNode]
+    });
+  }
+
+  registerObserver() {
+    if (!this._componentRef) {
+      return;
+    }
+    const node = getNodeById(this.context.ast, this.props.idyllASTNode.id);
+    let updated = false;
+    var observer = new MutationObserver(mutations => {
+      mutations.forEach(mutation => {
+        console.log('mutation', mutation);
+        updated = this.updateNode(
+          node,
+          mutation.oldValue,
+          mutation.target.data
+        );
+      });
+      if (updated) {
+        const ast = this.context.ast;
+        ast.id = getRandomId();
+        this.context.setAst(ast);
+      }
+    });
+    var config = {
+      subtree: true,
+      characterData: true,
+      characterDataOldValue: true
+    };
+    observer.observe(this._componentRef, config);
+    this.observer = observer;
+  }
+
+  componentDidMount() {
+    setCurrentCursorPosition(
+      this._componentRef,
+      cursorPositions[this.props.idyllASTNode.id]
+    );
+    // Set up mutation observer to catch DOM changes.
+    this.registerObserver();
+  }
+
+  handleMarkupRef(ref) {
+    if (!ref) {
+      return;
+    }
+    if (this.markupObserver) {
+      this.markupObserver.disconnect();
+    }
+    if (this.observer) {
+      this.observer.disconnect();
+    }
+    this._markupRef = ref;
+
+    // Set up mutation observer to catch DOM changes.
+    const node = getNodeById(this.context.ast, this.props.idyllASTNode.id);
+    var observer = new MutationObserver(mutations => {
+      mutations.forEach(mutation => {
+        const input = mutation.target.data;
+        this._markup = input;
+      });
+      // const ast = this.context.ast;
+      // ast.id = getRandomId();
+      // this.context.setAst(ast);
+    });
+    var config = {
+      subtree: true,
+      characterData: true,
+      characterDataOldValue: true
+    };
+    observer.observe(this._markupRef, config);
+    this.markupObserver = observer;
+  }
+
+  componentWillUnmount() {
+    this.observer && this.observer.disconnect();
+    this.markupObserver && this.markupObserver.disconnect();
+    this._componentRef
+      ? (cursorPositions[this.props.idyllASTNode.id] = getCurrentCursorPosition(
+          this._componentRef.parentNode
+        ))
+      : null;
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (
+      this.state.showCode === false &&
```

Comment:
`this.state.showCode === false` -> `!this.state.showCode`? Unless `showCode` isn't a boolean 😳

Reply:
Oof will update